mica/starcheck/starcheck.py

```diff
@@ -20,60 +22,197 @@
+    db = Ska.DBI.DBI(dbi='sqlite', server=config['server'])
+    # if we're outside of timelines, just try from the starcheck database
+    if date > last_tl or date < first_tl:
+        # Get one entry that is the last one before the specified time, in the most
+        # recently ingested products directory
+        starcheck = db.fetchone(
+            """select * from starcheck_obs, starcheck_id
+               where mp_starcat_time <= '{}' and mp_starcat_time > '{}'
+               and starcheck_id.id = starcheck_obs.sc_id
+               order by sc_id desc, mp_starcat_time desc """.format(
+                date, (DateTime(date) - 1).date))
+        if starcheck:
+            cat_info = get_starcheck_catalog(starcheck['obsid'], mp_dir=starcheck['dir'])
+            if date < first_tl:
+                cat_info['status'] = 'ran before timelines'
+            if date > last_tl:
+                cat_info['status'] = 'planned'
+            return cat_info
+    # If timelines apply, try to use those to find a commanded catalog
+    # Use kadi to get the first dwell that *ends* after the given time
+    dwells = events.dwells.filter(stop__gte=date, subset=slice(None, 1))
+    dwell = dwells[0]
+    # We want to search for legitimate commanding that would cover the time when a star
+    # catalog would have been commanded for this dwell.  This is generally the time range
+    # between the end of the previous dwell and the beginning of this dwell.  However, if
+    # there are multiple dwells from one maneuver, use the beginning of NMM from that one
+    # maneuver else, use the end of the last dwell.  Don't use nman_start time by default
+    # because that doesn't appear to work if the catalog was commanded in a nonstandard
+    # nmm sequence like dark cal.
+    start_cat_search = dwell.manvr.nman_start if dwell.manvr.n_dwell > 1 else dwell.get_previous().stop
+    timelines = timelines_db.fetchall(
+            """select * from timeline_loads where scs < 131
+           and datestop > '{}' and datestart < '{}' order by datestart""".format(
+            start_cat_search, dwell.start))
+    for timeline in timelines[::-1]:
+        starchecks = db.fetchall(
+            """select * from starcheck_obs, starcheck_id
+               where dir = '{}'
+               and mp_starcat_time >= '{}'
+               and mp_starcat_time <= '{}' and mp_starcat_time <= '{}'
+               and starcheck_id.id = starcheck_obs.sc_id
+               order by mp_starcat_time """.format(
+                timeline['mp_dir'],
+                timeline['datestart'],
+                timeline['datestop'], dwell.start))
+        # The last one should be the one before beginning of the dwell
+        if len(starchecks):
+            cat_info = get_starcheck_catalog(starchecks[-1]['obsid'], mp_dir=starchecks[-1]['dir'])
+            if cat_info['obs']['mp_starcat_time'] > DateTime().date:
+                cat_info['status'] = 'approved'
```

Comment:
Oops.  If the obsid it is in timelines but not in telemetry/kadi yet, this isn't going to work anyway (because we've already assumed finding a dwell earlier in this code).  I've forgotten, what's the fastest kadi way to get the last available dwell?